lib/ecto/changeset/relation.ex

```diff
@@ -83,11 +83,20 @@ defmodule Ecto.Changeset.Relation do
   def cast(%{cardinality: :many} = relation, params, current, on_cast) when is_map(params) do
     params =
       params
-      |> Enum.sort_by(&elem(&1, 0))
+      |> Enum.sort_by(&key_as_int/1)
       |> Enum.map(&elem(&1, 1))
     cast(relation, params, current, on_cast)
   end
 
+  defp key_as_int({key, _}) when is_binary(key) do
+    try do
+      String.to_integer(key)
```

Comment:
We should use `Integer.parse` and return the key unchanged if we can't parse it. That's because there is no requirement for string keys to represent integers.